pkg/resolver/multi.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package resolver
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"strings"
+)
+
+// MultiResolver errors.
+var (
+	ErrInvalidTLD         = errors.New("invalid TLD")
+	ErrResolverChainEmpty = errors.New("resolver chain empty")
+)
+
+type resolverMap map[string][]Interface
+
+// MultiResolver performs name resolutions based on the TLD of the URL.
+type MultiResolver struct {
+	resolvers resolverMap
+	// ForceDefault will force all names to be resolved by the default
+	// resolution chain, regadless of their TLD.
+	ForceDefault bool
+}
+
+// Option is a function that applies an option to a MultiResolver.
+type Option func(*MultiResolver)
+
+// NewMultiResolver will return a new MultiResolver instance.
+func NewMultiResolver(opts ...Option) *MultiResolver {
+	mr := &MultiResolver{
+		resolvers: make(resolverMap),
+	}
+
+	// Apply all options.
+	for _, o := range opts {
+		o(mr)
+	}
+
+	return mr
+}
+
+// WithForceDefault will force resolution using the default resolver
+// chain.
+func WithForceDefault() Option {
+	return func(mr *MultiResolver) {
+		mr.ForceDefault = true
+	}
+}
+
+// PushResolver will push a new Resolver to the name resolution chain for the
+// given TLD.
+// TLD names should be prepended with a dot (eg ".tld"). An empty TLD will push
+// to the default resolver chain.
+func (mr *MultiResolver) PushResolver(tld string, r Interface) error {
+	if tld != "" && !isTLD(tld) {
```

Comment:
I would allow empty tls value as the value for default resolvers. If there is no tld specified, to try to resolve with them.

Reply:
Empty value is specifically allowed. I've added a test with that specific case to confirm.